File: 589.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<int> preorder(Node* root) {
        if(root==NULL) return {};
        vector<int> result;
        stack<Node*> s;
        s.push(root);
        while(!s.empty())
        {
            Node* temp=s.top();
            s.pop();
            if(!temp->children.empty())
            {
                reverse(temp->children.begin(),temp->children.end());
                for(auto child: temp->children)
                {
                    s.push(child);
                }
            }
            result.push_back(temp->val);
        }
        return result;
    }
};

File: 589.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if root==None: return []
        result=[]
        stack=[]
        stack.append(root)
        while stack:
            temp=stack[-1]
            stack.pop();
            if temp.children:
                temp.children.reverse()
                for child in temp.children:
                    stack.append(child)
            result.append(temp.val)
        return result
